### DISCUSSION TITLE
How can I register a default resolver for a Type?

### DISCUSSION THREAD

**Discussion by steve-marmalade** [mirumee/ariadne#1201] [Q&A]
Consider the situation where I have a product type:

```graphql
type Query {
  topProducts: [Product]
  newProducts: [Product]
  ...
}

type Product {
    id: String!
    url: String!
    title: String!
    price: Float
    quantity: Int
    description: String
    ...
}
```

and let's say I have an API of the form:

```python
async def load_product(id: string) -> Product:
   ...
```

along with many different ways to get lists of product ids. My desired outcome is to write my "list" resolvers such they return lists of product ids:

```python
@query.field("topProducts")
async def resolve_top_products(_, info):
   top_product_ids = await fetch_top_product_ids()
  return [{"id": id_} for id_ in top_product_ids]
```

**And so here comes the feature request:**

Instead of writing something like the following for each product field:

```python
@product.field("title")
async def resolve_product_title(obj, info):
    if "title" not in obj:
        obj.update(await load_product(obj["id"]))
    return obj["title"]
```

I want to register a custom "special default resolver" (docs below) :

```python
@product.default
async def product_default_resolver(obj, info, field_name):
  if field_name not in obj:
    obj.update(await load_product(obj["id"]))
  return ariadne.default_resolver(obj, info, field_name)
```

Here are the [Ariadne docs for case conversion](https://ariadnegraphql.org/docs/case-conversion#setting-automatic-name-case-conversion-for-whole-schema):

> Types fields without resolver already set on them will be assigned a special resolver that seeks Python counterpart of camel case name in object's attributes or dicts keys. Eg. streetAddress2 field fill be resolved to street_address_2 attribute for objects and key for dicts.

So basically, I need a way to enhance this "special resolver" on a per-type basis. This would save me a lot of boilerplate.

**Comment by fedirz** [comment]
@steve-marmalade I'm looking into doing something similar. Have you figured out a workaround?

**Comment by pylipp** [comment]
@steve-marmalade I'm wondering whether you considered loading of products in the `topProducts` resolver?

```python
@query.field("topProducts")
async def resolve_top_products(_, info):
    top_product_ids = await fetch_top_product_ids()
    return [load_product(id_) for id_ in top_product_ids]
```

Assuming `load_product` returns an object with attributes `.title` or lookups like `["title"]` then the ariadne default resolver becomes effective and you don't need any custom resolvers.
